src/stactools/core/utils/raster_footprint.py

```diff
@@ -0,0 +1,301 @@
+from pystac import Item
+from pystac.utils import make_absolute_href
+from rasterio import Affine
+from rasterio.crs import CRS
+from rasterio.warp import transform_geom
+from shapely.geometry import mapping, shape
+from shapely.geometry.multipolygon import MultiPolygon
+from shapely.geometry.polygon import Polygon
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_SCALE = 1
+DEFAULT_PRECISION = 3
+
+
+def update_geometry_from_asset_footprint(
+    item: Item,
+    *,
+    asset_names: Optional[List[str]] = None,
+    scale: int = DEFAULT_SCALE,
+    precision: int = DEFAULT_PRECISION,
+    densification_factor: Optional[int] = None,
+    simplify_tolerance: Optional[float] = None,
+    no_data: Optional[int] = None,
+) -> Optional[Item]:
+    """
+    Accepts an Item and an optional list of asset names within that item, and updates
+    the geometry of that Item in-place with the data footprint derived from the first
+    of the assets that exists in the Item.
+
+    Two important operations during this calculation are the densification of the
+    footprint in the native CRS and simplification of the footprint after reprojection.
+    If the initial low-vertex polygon in the native CRS is not densified, this can result
+    in a reprojected polygon that does not accurately represent the data footprint. For
+    example, a MODIS asset represented by a rectangular 5 point Polygon in a sinusoidal
+    projection will reproject to a parallelogram in EPSG 4326, when it would be more
+    accurately represented by a polygon with two parallel sides and two curved sides.
+    The difference between these representations is even greater the further away from
+    the meridian and equator the asset is located.
+
+    After reprojection to EPSG 4326, the footprint may have more points than desired.
+    This can be simplified to a polygon with fewer points that maintain a minimum distance
+    between each point.
+
+    Args:
+        item (Item): The PySTAC Item to extend.
+        asset_names (Optional[List[str]]): The names of the assets for which to extract footprints.
+        scale (int): A factor to scale out the shape to. Defaults to 1.
+        precision (int): The number of decimal places to include in the coordinates for the
+            reprojected geometry. Defaults to 3 decimal places.
+        densification_factor (Optional[int]): The factor by which to increase point density within
+            the polygon.
+            A factor of 2 would double the density of points (placing one new point between each
+            existing pair
+            of points), a factor of 3 would place two points between each point, etc.
+        simplify_tolerance (Optional[float]): All points in the simplified object will be within
+            the tolerance distance of the original geometry, in degrees.
+
+    Returns:
+        Iterator[Tuple[str, Dict[str, Any]]]: Iterator of the data extent as a geojson dict
+            for each asset.
```

Comment:
```suggestion
       Optional[pystac.Item]: The modified item, or ``None`` if the item was unmodified.
```

It feels a little awkward to have a "partially-fluent" interface, where you can't rely on the item being returned. IMO we should either
- always return the item, or
- return a boolean indicating whether the item was modified

Reply:
yeah, I agree. changed.